routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,187 @@
+  virtual TUnpackedPathSegments const & GetSegments() const override { return m_pathSegments; }
+
+  virtual void GetPossibleTurns(TNodeId node, m2::PointD const & ingoingPoint,
+                                m2::PointD const & junctionPoint, size_t & ingoingCount,
+                                TTurnCandidates & outgoingTurns) const override
+  {
+    ingoingCount = 0;
+    outgoingTurns.clear();
+
+    if (node >= m_routeEdges.size())
+    {
+      ASSERT(false, (m_routeEdges.size()));
+      return;
+    }
+
+    auto adjacentEdges = m_adjacentEdges.find(node);
+    if (adjacentEdges == m_adjacentEdges.cend())
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    ingoingCount = adjacentEdges->second.m_ingoingTurnsCount;
+    outgoingTurns = adjacentEdges->second.m_outgoingTurns;
+  }
+
+  virtual double GetPathLength() const override { return m_routeLength; }
+
+  virtual m2::PointD const & GetStartPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.front().GetStartJunction().GetPoint();
+  }
+
+  virtual m2::PointD const & GetEndPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.back().GetEndJunction().GetPoint();
+  }
+
+private:
+  IRoadGraph::TEdgeVector const & m_routeEdges;
+  AdjacentEdgesMap const & m_adjacentEdges;
+  TUnpackedPathSegments const & m_pathSegments;
+  double m_routeLength;
+};
+
+ftypes::HighwayClass GetHighwayClass(FeatureID const & featureId, Index const & index)
+{
+  ftypes::HighwayClass highWayClass = ftypes::HighwayClass::Undefined;
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  uint32_t const featureIndex = featureId.m_index;
+
+  FeatureType ft;
+  Index::FeaturesLoaderGuard loader(index, mwmId);
+  loader.GetFeatureByIndex(featureIndex, ft);
+  highWayClass = ftypes::GetHighwayClass(ft);
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Error, ());
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Undefined, ());
+  return highWayClass;
+}
+}  // namespace
+
+namespace routing
+{
+BicycleDirectionsEngine::BicycleDirectionsEngine(Index const & index) : m_index(index) {}
+
+void BicycleDirectionsEngine::Generate(IRoadGraph const & graph, vector<Junction> const & path,
+                                       Route::TTimes & times, Route::TTurns & turns,
+                                       vector<m2::PointD> & routeGeometry,
+                                       my::Cancellable const & cancellable)
+{
+  size_t const pathSize = path.size();
+  CHECK_NOT_EQUAL(pathSize, 0, ());
+
+  times.clear();
+  turns.clear();
+  routeGeometry.clear();
+  m_adjacentEdges.clear();
+  m_pathSegments.clear();
+
+  auto emptyPathWorkaround = [&]()
+  {
+    turns.emplace_back(pathSize - 1, turns::TurnDirection::ReachedYourDestination);
+    this->m_adjacentEdges[0] = AdjacentEdges(1);  // There's one ingoing edge to the finish.
+  };
+
+  if (pathSize <= 1)
+  {
+    ASSERT(false, (pathSize));
```

Comment:
I got ready the code for path of 0-1 points and added some integration tests on it.